_posts/not-all-changes-are-born-equal.md

```diff
@@ -0,0 +1,71 @@
+tags:
+- Design
+- Architecture
+- SOA
+- Coupling
+---
+
+I recently gave an online talk titled ["Long live and prosper"](https://milestone.topics.it/talks/long-live-and-prosper.html). The conference session is in Italian and available on [YouTube](https://www.youtube.com/watch?v=YOJ0rQ2gn44). Unfortunately, the only recording of this conference session is in Italian; hopefully, I'll deliver the English version again soon and get the recording published.
+
+During the usual after talk Q&A session, I had the opportunity to answer questions, among which the following two gave me the idea to write this post:
+
+- Shouldn't a Microservices/Distributed architecture facilitate the ability of the system to change?
+- Is it possible that a new business vision invalidates service boundaries?
+
+> Original questions were in Italian; I took the liberty of translating them.
+
+The short answer, a telegraphic one, to both questions is: yes. The longer answer requires a bit of history and requires starting from a different point of view.
+
+## Coupling
+
+We're terrified by coupling. We spend countless hours discussing how bad it is. And in the end, we risk overengineering the design and the solutions we put in place.
+
+To begin with, why are we terrified by coupling?
+
+Historically we've been brainwashed that "coupling is bad." We've probably been bitten by coupling once or twice in our career, leading a project to end badly or fail. And once burned a couple of times, it's easy to become overcautious.
+
+Let's step back for a second and resonate on coupling. First, zero coupling is impossible. Zero coupling is one of those utopic goals not that different from a completely secure computer. An utterly safe computer is not connected to anything, including a human being. A zero coupling architecture is an architecture that does nothing; there is no connection between all the parts composing it.
+
+When designing a system, the goal is to reduce coupling where it makes sense, accept coupling where it makes sense, encourage coupling where beneficial, and do not try to fight coupling where there is nothing to fight.
+
+### Hold on, can coupling be beneficial?
+
+Yes, it can. Let's use a logging library as a sample. If we're designing a component to be used by others, we need to allow users to plugin the logging library they want. In this situation, the traditional approach is to rely on an interface, something like `ILogger`. The component we're designing depends on `ILogger`, and the user, when using it, will provide an `ILogger` implementation. This approach decouples our code from the logging library implementation details, and users can provide their logging library, whose only requirement is to respect the interface contract.
```

Comment:
```suggestion
Yes, it can. Let's use a logging library as an example. If we're designing a component to be used by others, we need to allow users to plug in the logging library they want. In this situation, the traditional approach is to rely on an interface, something like `ILogger`. The component we're designing depends on `ILogger`, and the user, when using it, will provide an `ILogger` implementation. This approach decouples our code from the logging library implementation details, and users can provide their logging library, whose only requirement is to respect the interface contract.
```